ariadne/src/Ariadne/Wallet/Backend/KeyStorage.hs

```diff
@@ -137,170 +133,217 @@ runCatchInState m = StateT $ \s ->
     Left e -> (Left e, s)
     Right (a, s') -> (Right a, s')
 
--- | Get the wallet index by name or using current selection.
+-- | Get the wallet HdRootId by name, ID, UI index or using current selection.
 resolveWalletRef
   :: IORef (Maybe WalletSelection)
-  -> (CardanoMode ~> IO)
   -> WalletReference
-  -> IO Word
-resolveWalletRef walletSelRef runCardanoMode = \case
+  -> DB
+  -> IO HdRootId
+resolveWalletRef walletSelRef walRef db = case walRef of
   WalletRefSelection -> do
     mWalletSelection <- readIORef walletSelRef
     case mWalletSelection of
       Nothing -> throwM NoWalletSelection
-      Just WalletSelection{..} -> return wsWalletIndex
-  WalletRefByName (WalletName name) -> do
-    us <- runCardanoMode getSecretDefault
-    case findIndex (\w -> w ^. wdName == name) (us ^. usWallets) of
-      Just i -> return (fromIntegral i)
-      Nothing -> throwM $ WalletDoesNotExist name
-  WalletRefByIndex i -> return i
-
--- | Like 'resolveWalletRef', but for accounts.
+      Just selection -> return $ getHdRootId selection
+  WalletRefByUIindex i -> do
+    -- Note: Add/remove wallets cause changes in indexation
+    case walletList ^? ix (fromIntegral i) of
+      Just hdRoot -> return (hdRoot ^. hdRootId)
+      Nothing -> throwM $ WalletIndexOutOfRange i
+  WalletRefByHdRootId hrR -> case readHdRoot hrR hdWallets of
+    Right _ -> return hrR
+    Left err -> throwM err
+
+  where
+    hdWallets :: HdWallets
+    hdWallets = db ^. dbHdWallets
+
+    getHdRootId :: WalletSelection -> HdRootId
+    getHdRootId (WSRoot rId) = rId
+    getHdRootId (WSAccount acId) = acId ^. hdAccountIdParent
+
+    walletList :: [HdRoot]
+    walletList = toList (readAllHdRoots hdWallets)
+
+  -- TODO:
+  -- * Check if is it possible, that address or account exist in db
+  --   but has no parents?
+  -- * If so, where do we need to check,
+  --   if account and address parsents exist?
 resolveAccountRef
   :: IORef (Maybe WalletSelection)
-  -> (CardanoMode ~> IO)
   -> AccountReference
-  -> IO (Word, Word32)
-resolveAccountRef walletSelRef runCardanoMode accountRef = do
-    walletIdx <- resolveWalletRef walletSelRef runCardanoMode walletRef
-    (walletIdx, ) <$> resolveAccountIdx walletIdx
+  -> DB
+  -> IO HdAccountId
+resolveAccountRef walletSelRef accountRef walletDb = case accountRef of
+    AccountRefSelection -> do
+      mWalletSelection <- readIORef walletSelRef
+      case mWalletSelection of
+        Nothing -> throwM NoWalletSelection
+        Just (WSRoot _) -> throwM NoAccountSelection
+        Just (WSAccount accId) -> do
+            checkParentRoot accId
+            return accId
+    AccountRefByHdAccountId accId -> do
+      checkParentRoot accId
+      return accId
+    AccountRefByUIindex accIdx walRef -> do
+      accounts <- getAccList walRef
+      acc <- maybeThrow
+        (AccountDoesNotExist $ pretty accIdx)
```

Comment:
I want errors to be consistent, so it should be `AccountIndexOutOfRange` akin to `WalletIndexOutOfRange`, I suppose.